app/com/arpnetworking/rollups/ConsistencyChecker.java

```diff
@@ -0,0 +1,579 @@
+        final Optional<Throwable> failure = sampleCounts.getFailure();
+
+        final Metrics metrics = _metricsFactory.create();
+        metrics.addAnnotation("trigger", task.getTrigger().name());
+        metrics.addAnnotation("period", task.getPeriod().name());
+
+        // TODO(spencerpearson): it might be useful to see how behavior varies by cardinality, something like
+        // metrics.addAnnotation("log_realized_cardinality", Long.toString((long) Math.floor(Math.log10(task.getRealizedCardinality()))));
+
+        metrics.incrementCounter("rollup/consistency_checker/query_successful", failure.isPresent() ? 0 : 1);
+
+        if (failure.isPresent()) {
+            LOGGER.warn()
+                    .setMessage("failed to query Kairos for sample-counts")
+                    .addData("task", task)
+                    .setThrowable(failure.get())
+                    .log();
+            return;
+        }
+
+        final double nOriginalSamples = sampleCounts.getSourceSampleCount();
+        final double nSamplesDropped = nOriginalSamples - sampleCounts.getRollupSampleCount();
+        final double fractionalDataLoss = nSamplesDropped / nOriginalSamples;
+        metrics.setGauge("rollup/consistency_checker/fractional_data_loss", fractionalDataLoss);
+        final LogBuilder logBuilder =
+                // TODO(spencerpearson): probably make this level-thresholding configurable?
+                fractionalDataLoss == 0 ? LOGGER.trace()
+                : fractionalDataLoss < 0.001 ? LOGGER.debug()
+                : fractionalDataLoss < 0.01 ? LOGGER.info()
+                : fractionalDataLoss < 0.1 ? LOGGER.warn()
+                : LOGGER.error();
+        logBuilder.setMessage((fractionalDataLoss == 0 ? "no " : "") + "data lost in rollup")
+                .addData("task", task)
+                .addData("nOriginalSamples", nOriginalSamples)
+                .addData("nSamplesDropped", nSamplesDropped)
+                .addData("fractionalDataLoss", fractionalDataLoss)
+                .log();
+
+        /* TODO(spencerpearson, OBS-1176): re-trigger re-execution of bad datapoints, something like
+
+            if (discrepancyTooBig) {
+                _rollupManagerPool.tell(
+                    new RollupDefinition.Builder()
+                            .setSourceMetricName(task.getSourceMetricName())
+                            .setDestinationMetricName(task.getRollupMetricName())
+                            .setStartTime(task.getStartTime())
+                            .setPeriod(task.getPeriod())
+                            .build()
+                );
+            }
+
+         */
+    }
+
+    /* package private */ static SampleCounts parseSampleCounts(
+            final Task task,
+            final MetricsQueryResponse response
+    ) throws MalformedSampleCountResponse {
+        final Map<String, Long> countsByMetric = Maps.newHashMap();
+        if (response.getQueries().size() != 2) {
```

Comment:
Can you add a comment to explain why you look for size 2 but emit an exception saying you needed exactly 1?

Reply:
Copy-paste error! Fixed.